Move Ada packages into own namespace

heijligen opened the pull request:
## Motivation
Ada dependencies need to be build with the same compiler as the main project. There is currently no possibility to use a specific gnat version and the prebuild packages.

Also spark2014 (gnatprove) is currently only available in version `fsf-12`. To build the newer `fsf-13` version all ada dependencies musst be build with gnat-13.

## Description of changes
- Move all ada packages into ada-modules
- Add the `gnatPackages` namespace for ada packages
- Rename `spark2014` to `gnatprove` (same as AdaCores alire [package](https://alire.ada.dev/crates/gnatprove.html))
- Make the `src` of `gnatprove` dependent on the gnat version
- Add `gnat12Packages` and `gnat13Packages` as version specific versions

## Options
- I'm open to become maintainer of the ada stuff

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

heijligen commented on pkgs/development/ada-modules/gnatprove/default.nix:
```diff
@@ -1,29 +1,46 @@
+stdenv.mkDerivation rec {
+  pname = "gnatprove";
+
+  version = "fsf-${gnat_version}";
```
My intention was to have the branch name in the package name to make clear on which gcc version this gnatprove is based.
For the last 3 versions (11.2, 12, 13) there where no additional commits later on that branch.

What do you think of `gnatprove-fsf-13_12db22e854`? So including the short hash of that commit?
Or `gnatprove-fsf-12_2022-05-25` with the date?

jiegec replied:
The date is better IMO.

heijligen replied:
I've updated the commits to use `gnatprove-fsf-12_2022-05-25` as version scheme.

jiegec commented:
Result of `nixpkgs-review pr 283123` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>44 packages built:</summary>
  <ul>
    <li>gnatPackages.gnat (gnat12Packages.gnat)</li>
    <li>gnatPackages.gnat.info (gnat12Packages.gnat.info)</li>
    <li>gnatPackages.gnat.man (gnat12Packages.gnat.man)</li>
    <li>gnatPackages.gnatcoll-core (gnat12Packages.gnatcoll-core)</li>
    <li>gnatPackages.gnatcoll-db2ada (gnat12Packages.gnatcoll-db2ada)</li>
    <li>gnatPackages.gnatcoll-gmp (gnat12Packages.gnatcoll-gmp)</li>
    <li>gnatPackages.gnatcoll-iconv (gnat12Packages.gnatcoll-iconv)</li>
    <li>gnatPackages.gnatcoll-lzma (gnat12Packages.gnatcoll-lzma)</li>
    <li>gnatPackages.gnatcoll-omp (gnat12Packages.gnatcoll-omp)</li>
    <li>gnatPackages.gnatcoll-postgres (gnat12Packages.gnatcoll-postgres)</li>
    <li>gnatPackages.gnatcoll-python3 (gnat12Packages.gnatcoll-python3)</li>
    <li>gnatPackages.gnatcoll-readline (gnat12Packages.gnatcoll-readline)</li>
    <li>gnatPackages.gnatcoll-sql (gnat12Packages.gnatcoll-sql)</li>
    <li>gnatPackages.gnatcoll-sqlite (gnat12Packages.gnatcoll-sqlite)</li>
    <li>gnatPackages.gnatcoll-syslog (gnat12Packages.gnatcoll-syslog)</li>
    <li>gnatPackages.gnatcoll-xref (gnat12Packages.gnatcoll-xref)</li>
    <li>gnatPackages.gnatcoll-zlib (gnat12Packages.gnatcoll-zlib)</li>
    <li>gnatPackages.gnatinspect (gnat12Packages.gnatinspect)</li>
    <li>gnatprove (gnatPackages.gnatprove)</li>
    <li>gnatPackages.gprbuild (gnat12Packages.gprbuild)</li>
    <li>gnatPackages.gprbuild-boot (gnat12Packages.gprbuild-boot)</li>
    <li>gnatPackages.xmlada (gnat12Packages.xmlada)</li>
    <li>gnat13Packages.gnat</li>
    <li>gnat13Packages.gnat.info</li>
    <li>gnat13Packages.gnat.man</li>
    <li>gnat13Packages.gnatcoll-core</li>
    <li>gnat13Packages.gnatcoll-db2ada</li>
    <li>gnat13Packages.gnatcoll-gmp</li>
    <li>gnat13Packages.gnatcoll-iconv</li>
    <li>gnat13Packages.gnatcoll-lzma</li>
    <li>gnat13Packages.gnatcoll-omp</li>
    <li>gnat13Packages.gnatcoll-postgres</li>
    <li>gnat13Packages.gnatcoll-python3</li>
    <li>gnat13Packages.gnatcoll-readline</li>
    <li>gnat13Packages.gnatcoll-sql</li>
    <li>gnat13Packages.gnatcoll-sqlite</li>
    <li>gnat13Packages.gnatcoll-syslog</li>
    <li>gnat13Packages.gnatcoll-xref</li>
    <li>gnat13Packages.gnatcoll-zlib</li>
    <li>gnat13Packages.gnatinspect</li>
    <li>gnat13Packages.gnatprove</li>
    <li>gnat13Packages.gprbuild</li>
    <li>gnat13Packages.gprbuild-boot</li>
    <li>gnat13Packages.xmlada</li>
  </ul>
</details>

heijligen commented:
@sternenseemann: Would be nice if you could review this PR. Thanks

sternenseemann commented:
@heijligen I'll try to do so soon! After an initial look I'd prefer if `lib.makeScopeWithSplicing` would be used. `newScope` isn't a huge issue because we can't really build a functioning Ada cross-compiler, however it would be good if `gprbuild` which can also be used for other languages would splice correctly when used in a cross situation. Or was there a specific problem that made you choose `newScope`?

The advantage is also that we don't have to break any interfaces if we manage to get cross to work.

heijligen commented:
I've found `newScope` beeing used in other `*Packages` like `ocamlPackages`. `lib.makeScopeWithSplicing` is new to me. I'll have a look and try it.

Thanks

heijligen commented:
I couldn't figure out how to use `lib.makeScopeWithSplicing`. Is there any documentation?

sternenseemann commented:
Not to my knowledge unfortunately except existing uses in Nixpkgs… I suppose it doesn't necessarily make sense if you're unfamiliar with the internals of the cross compilation system.

heijligen commented:
So should I get into cross compiling to get this upstream or is `newScope` ok until gnat can be used for cross builds?

heijligen committed:
gnatprove: Rename spark2014

Gnatprove is the tool distributed out of the spark2014 repository.
AdaCore, the developer behind it, distribute it under the name gnatprove
in the Ada alire package repository [0].
Rename the package to keep in line with other distribution channels.

[0] https://alire.ada.dev/crates/gnatprove

heijligen committed:
gnatprove: Make src dependend on gcc version

Gnatprove is tightly bound to a specific gcc version. Use the gcc input
version to determin which gnatprove/spark2014 sources should be used.
For now gcc 12 is the only working option. gcc 13 will work as soon as
all ada buildInputs can be provided as gcc 13 build.

heijligen committed:
gnatPackages: Add scope for all ada packages

Ada depencencies musst be build with the same gnat version as the
project. Use a namespace as preperation to build with different gnat
versions.

gprbuild and gnatprove are still globaly visable.

heijligen committed:
gnatPackages: Build all ada-modules with gnat12 and gnat13

Add namespace gnat12Packages and gnat13Packages to get ada-modules build
with a specific gnat version.
For gnatprove this means building the gnat12 and gnat13 version.

heijligen commented:
@sternenseemann I've updated this to my best knowledge with `lib.makeScopeWithSplicing'`

heijligen committed:
gnatprove: Expose SPARKlib through gpr

SPARKlib is shipped with gnatprove / spark2014 but the gpr files are in
lib/gnat/. Nix gprbuild cant pick them up there. Link them to a location
where gprbuild can find them.

sternenseemann reviewed (changes requested):
* The changes need an entry in the changelog for 24.05 (under breaking changes), describing what's changed and what existing users of Ada in nixpkgs need to consider.
* To make the transition a bit easier, add aliases to `pkgs/top-level/aliases.nix` for backwards compatibility for external users (these aliases can't be used in nixpkgs). They'll be automatically removed after a grace period.
  * spark2014 (probably with a `throw` that says that it has been removed?)
  * gnatcoll-db2ada
  * gnatinspect
  * gprbuild-boot
  * gprbuild
  * gnatcoll-core
  * gnatcoll-gmp
  * gnatcoll-iconv
  * gnatcoll-lzma
  * gnatcoll-omp
  * gnactoll-python3
  * gnatcoll-readline
  * gnatcoll-syslog
  * gnatcoll-zlib
  * gnatcoll-postgres
  * gnatcoll-sql
  * gnatcoll-sqlite
  * gnatcoll-xref
  * xmlada

sternenseemann committed:
gprbuild: prevent cross compiled gprbuild-boot being pulled in

Building a cross-compiled gprbuild-boot would obviously not succeed.
This change is kind of academic since we don't really have a working Ada
cross-compiler in nixpkgs, but helps verifying changes to the
gnatPackages infrastructure.

heijligen commented on pkgs/development/ada-modules/gnatprove/default.nix:
```diff
@@ -1,29 +1,49 @@
 , makeWrapper
 }:
+let
+  gnat_version = builtins.substring 0 2 "${gnat.version}";
```
`lib.version.major gnat.version` returns `error: value is a string while a set was expected`, so this isn't working

heijligen replied:
Sorry for the noise.
There was a spelling error leading to the error message: `version` -> `versions` as below in the comment.

sternenseemann replied:
Yeah, lib.versions, sorry.

heijligen committed:
release-notes: Add Ada / gnatPackages changes

heijligen commented on pkgs/top-level/all-packages.nix:
```diff
@@ -16096,6 +16096,14 @@ with pkgs;
     bintools = bintoolsDualAs;
   });
 
+  gnatPackages = recurseIntoAttrs (callPackage ./ada-packages.nix { });
```
I've intentionally used the default gnat there. So `gnat` and `gnatPackages` (the one without a version) are always in sync

heijligen commented:
> * The changes need an entry in the changelog for 24.05 (under breaking changes), describing what's changed and what existing users of Ada in nixpkgs need to consider.

I've added an entry in the changelog in a own commit. Is this enough or should it be longer

> * To make the transition a bit easier, add aliases to `pkgs/top-level/aliases.nix` for backwards compatibility for external users (these aliases can't be used in nixpkgs). They'll be automatically removed after a grace period.

Done

>   
>   * spark2014 (probably with a `throw` that says that it has been removed?)

spark2014 is just renamed. I've added an alias. 
gnatprove (spark2014) and gprbuild are still visible in the top-level scope. Therefore I've not added an alias to gnatPackages.<pname>

heijligen commented on pkgs/top-level/ada-packages.nix:
```diff
@@ -0,0 +1,46 @@
+{ lib
+, pkgs
+, config
```
Done

heijligen commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17947,10 +17955,6 @@ with pkgs;
 
   sparkleshare = callPackage ../applications/version-management/sparkleshare { };
 
-  spark2014 = callPackage ../development/libraries/ada/spark2014 {
```
It should be gone.
At least `gs/development/libraries/ada/spark2014/default.nix` is deleted.

sternenseemann commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -173,6 +173,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - The `cudaPackages` package scope has been updated to `cudaPackages_12`.
 
+- Ada packages (libraries and tools) are now in the gnatPackages scope. gnatPackages uses the default gnat compiler, gnat12Packackages and gnat13Packages uses the respective compiler version. Gnatprove is build in the version of the compiler.
```
Minor editing I'd do:

```suggestion
- Ada packages (libraries and tools) have been moved into the `gnatPackages` scope. `gnatPackages` uses the default GNAT compiler, `gnat12Packackages` and `gnat13Packages` use the respective matching compiler version.
```

For spark2014 I'd write something like »`spark2014` has been renamed to `gnatprove`. A version of `gnatprove` matching different GNAT versions is available from the different `gnatPackages` sets.«

heijligen replied:
Done

sternenseemann commented on pkgs/top-level/all-packages.nix:
```diff
@@ -16075,6 +16075,13 @@ with pkgs;
 
+  gnat12Packages = recurseIntoAttrs (callPackage ./ada-packages.nix { gnat = buildPackages.gnat12; });
+  gnat13Packages = recurseIntoAttrs (callPackage ./ada-packages.nix { gnat = buildPackages.gnat13; });
+  gnatPackages   = recurseIntoAttrs (callPackage ./ada-packages.nix { gnat = buildPackages.gnat; });
```
Would still be nice to make this an alias to avoid duplicate eval work. Maybe move it next to the binding for `gnat = `, so you can't forget to update it?

heijligen replied:
Ack, I've added a comment behind `gnat=gnat12` to also update `gnatPackages`.

heijligen commented:
@sternenseemann Thanks for for all your work in reviewing and helping me with this PR!
